Add velbus HVAC options

RZomermanMS opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

RZomermanMS committed:
Added HVAC options

cereal2nd reviewed (changes requested):
This will also require a new version of velbusaio.

In the lib was also a change needed

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Update manifest.json

required aio to 2023.12.0

joostlek commented on homeassistant/components/velbus/manifest.json:
```diff
@@ -13,7 +13,7 @@
     "velbus-protocol"
   ],
-  "requirements": ["velbus-aio==2023.11.0"],
+  "requirements": ["velbus-aio==2023.12.0"],
```
Please bump in a separate PR

joostlek replied:
^

RZomermanMS replied:
the bump will be required for the new code base to work..

joostlek replied:
Yes, but the code isn't required for the bump to work. So I would love to see a different PR with the bump and then we can rebase this on the latest version and this keeps the PRs pure, small and clean.

RZomermanMS replied:
separate dependency bump was already accepted and merged: https://github.com/home-assistant/core/pull/107066

RZomermanMS replied:
@joostlek  - would you be able to review?

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Update manifest.json

RZomermanMS committed:
Add files via upload

RZomermanMS committed:
Merge branch 'home-assistant:dev' into dev

Cereal2nd commented:
once this one is merged we can approve this PR

https://github.com/home-assistant/core/pull/107066

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS reviewed (commented):
should all be good now - as dependency is already on   "requirements": ["velbus-aio==2023.12.0"],

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Merge branch 'dev' into dev

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -18,6 +18,11 @@
 from .entity import VelbusEntity, api_call
 
+THERMOSTAT_MODE_MAP: dict[str, HVACMode] = {
+    """Set up Velbus switch based on config_entry.""" "heat": HVACMode.HEAT,
```
Something went wrong with a docstring here

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -78,3 +92,11 @@ async def async_set_preset_mode(self, preset_mode: str) -> None:
+    @api_call
+    async def async_set_hvac_mode(self, **kwargs: str) -> None:
+        """Set the new hvac mode."""
+        if (mode := kwargs.get("hvac_mode")) is None:
+            return
+        await self._channel.set_mode(mode)
```
What's the type of `mode` here?

RZomermanMS replied:
heating or cooling.. 

The Velbus bus system reports back the actual mode (heating / cooling) in the bus protocol.. we take that actual mode and set it in HA (also to accommodate switches done externally to HA on the mode)  and if switched in HA - we send the command on the bus to switch the actual device to that mode.

joostlek replied:
I mean, is it `cool`, `heat`, `True` or `False`

RZomermanMS replied:
ah, indeed "cool" or "heat"

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -65,6 +69,16 @@ def current_temperature(self) -> int | None:
+    @property
+    def hvac_mode(self) -> HVACMode:
+        """Return the current hvac mode based on cool_mode message."""
+        if (mode := self._channel.get_cool_mode()) is None:
+            return
+        elif mode is True:
+            return HVACMode.COOL
+        elif mode is False:
+            return HVACMode.HEAT
```
```suggestion
        return {
            True: HVACMode.COOL,
            False: HVACMode.HEAT,
        }.get(self._channel.get_cool_mode())
```

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -18,6 +18,11 @@
 from .const import DOMAIN, PRESET_MODES
 from .entity import VelbusEntity, api_call
 
+THERMOSTAT_MODE_MAP: dict[str, HVACMode] = {
+    """Set up Velbus switch based on config_entry.""" "heat": HVACMode.HEAT,
+    "cool": HVACMode.COOL,
+}
+THERMOSTAT_INV_MODE_MAP = {v: k for k, v in THERMOSTAT_MODE_MAP.items()}
```
You're not using them at the moment. Also because HVACMode.COOL already means "cool" so there is no real need to to a conversion for the Velbus lib to understand what you mean.

Be aware, the moment that Velbus introduces HVAC modes that are not 1:1 mapped, we need this again

RZomermanMS committed:
Update homeassistant/components/velbus/climate.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

RZomermanMS commented:
still checking

emontnemery commented:
@RZomermanMS you've marked the pull request "Ready for review", but it seems you have not addressed Joost's review comments. Can you please double check?

emontnemery commented:
(Close + Open to try to retrigger stuck CI)

RZomermanMS committed:
Update climate.py

removed unused variables for cool and heat

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS commented:
> @RZomermanMS you've marked the pull request "Ready for review", but it seems you have not addressed Joost's review comments. Can you please double check?

yes. addressed the comments.. should all be ok now

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -19,6 +19,9 @@
 from .const import DOMAIN, PRESET_MODES
 from .entity import VelbusEntity, api_call
 
+THERMOSTAT_MODE_MAP: dict[str, HVACMode] = {}
+
+THERMOSTAT_INV_MODE_MAP = {v: k for k, v in THERMOSTAT_MODE_MAP.items()}
```
Where do we now use this?

RZomermanMS replied:
that I dont know.. its part of the old code.. but removing line 22 renders the controls unusable.. and the thermostat on the HA portal stays greyed out.. I believe it is used to read the possible options from velbusIO and push those to HA.- as well as copy the controls from HA to velbusIO

joostlek replied:
But there's nothing in home assistant that uses this and the code you added don't use this, so I'm not sure how that happens

RZomermanMS replied:
hmm you might be right.. I've dropped both lines.. and it seems to continue to work..

RZomermanMS committed:
Update climate.py

removed unused functions

emontnemery commented on homeassistant/components/velbus/climate.py:
```diff
@@ -66,6 +64,14 @@ def current_temperature(self) -> int | None:
+    @property
+    def hvac_mode(self) -> HVACMode:
+        """Return the current hvac mode based on cool_mode message."""
+        return {
+            True: HVACMode.COOL,
+            False: HVACMode.HEAT,
+        }.get(self._channel.get_cool_mode())
```
Please move the map to a module level constant:

```suggestion
        # Map velbus cool mode to Home Assistant HVACMode
        COOL_MODE_MAP = {
            True: HVACMode.COOL,
            False: HVACMode.HEAT,
        }
        
        ...

        return COOL_MODE_MAP.get(self._channel.get_cool_mode())
```

emontnemery commented on homeassistant/components/velbus/climate.py:
```diff
@@ -79,3 +85,11 @@ async def async_set_preset_mode(self, preset_mode: str) -> None:
+    @api_call
+    async def async_set_hvac_mode(self, **kwargs: str) -> None:
+        """Set the new hvac mode."""
+        if (mode := kwargs.get("hvac_mode")) is None:
+            return
```
This is impossible, the service schema requires `hvac_mode` to be a valid `HVACMode`.
OTOH, `HVACMode` allows other modes than `cool` and `heat` and I think we need to guard against *that*.

RZomermanMS replied:
removed the if none ..

emontnemery commented on homeassistant/components/velbus/climate.py:
```diff
@@ -79,3 +85,11 @@ async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the new hvac mode."""
+        if (mode := kwargs.get("hvac_mode")) is None:
+            return
+        await self._channel.set_mode(mode)
```
Add a map mapping from Home Assistant HVACMode to strings "cool" and "heat" supported by velbus.
Be sure to catch `KeyError`, and translate that to `ServiceValidationError` to address my previous comment about unsupported modes.

RZomermanMS replied:
remove the none state

emontnemery commented on homeassistant/components/velbus/climate.py:
```diff
@@ -66,6 +64,14 @@ def current_temperature(self) -> int | None:
+        return {
+            True: HVACMode.COOL,
+            False: HVACMode.HEAT,
+        }.get(self._channel.get_cool_mode())
```
Based on a quick glance at the library, `get_cool_mode` may raise an attribute error if we didn't get a `cool_mode` from the device. Either update the library to initialize `_cool_mode` or handle the attribute error here.

RZomermanMS replied:
can you elaborate on this a bit ?

emontnemery replied:
This is the implementation in velbusaio: https://github.com/Cereal2nd/velbus-aio/blob/master/velbusaio/channels.py#L551-L552

The problem is that `self._cool_mode` is not guaranteed to be set, and if it isn't, calling get_cool_mode() while cause an attributeerror to be raised.

RZomermanMS replied:
Do I need to make a change in velbusIO then to match: 

    def get_cool_mode(self) -> str:
        if self is none;
           return none
        return self._cool_mode
        
        ?

emontnemery replied:
In velbusIO the fix would maybe be like this:
```diff
 class Temperature(Channel):
     """
     A Temperature sensor channel
     """
 
     _cur = 0
     _cur_precision = None
     _max = None
     _min = None
     _target = 0
     _cmode = None
+    _coolmode = None
     _cstatus = None
     _thermostat = False
     _sleep_timer = 0
```

RZomermanMS committed:
Update homeassistant/components/velbus/climate.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

RZomermanMS committed:
Update climate.py

accepted changes

RZomermanMS committed:
Update climate.py

remove state None for HVAC-MODE

RZomermanMS committed:
Update climate.py

changed set_hvac_mode to remove none and only switch when state /= requested mode

RZomermanMS committed:
Update climate.py

indent on line 94/95

RZomermanMS committed:
Update climate.py

changed set_hvac_mode attribute type to match superclass ClimateEntity (HVACMode)

RZomermanMS committed:
Update climate.py

changed def hvac_mode to 2 return options (to avoid any)

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Update climate.py

ruff formatting

emontnemery commented on homeassistant/components/velbus/climate.py:
```diff
@@ -79,3 +83,10 @@ async def async_set_preset_mode(self, preset_mode: str) -> None:
+    @api_call
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set the new hvac mode."""
+        if hvac_mode not in self.hvac_mode:
+            await self._channel.set_mode(hvac_mode)
+            self.async_write_ha_state()
```
This makes no sense, please fix it

RZomermanMS replied:
This essentially says - if the current mode is not the same as the requested mode - switch to the requested mode.. if already in requested mode.. dont do anything..

emontnemery replied:
"not in" is not an idiomatic way to test for string unequality in Python.

I think you misunderstood my comment here: <https://github.com/home-assistant/core/pull/106570#discussion_r1627170542>

What I meant was that it's possible for the user to pass an `hvac_mode` which is not in `hvac_modes`. That should be detected by something like this:

```py
        if hvac_mode not in self.hvac_modes:
            raise ServiceValidationError(
                translation_domain=DOMAIN,
                translation_key=unsupported_hvac_mode
                translation_placeholders={"hvac_mode":hvac_mode}
```

Cereal2nd commented:
@RZomermanMS  can we look at these last commetns so we can get this one merged?

RZomermanMS committed:
Update climate.py

added serviceValidationError section in hvac_mode setting

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Update climate.py

RZomermanMS committed:
Merge branch 'dev' into dev

RZomermanMS committed:
Update climate.py

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -79,3 +85,16 @@ async def async_set_preset_mode(self, preset_mode: str) -> None:
+        if hvac_mode not in self._attr_hvac_modes:
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="invalid_hvac_mode",
```
This requires an exception string in the `strings.json`

RZomermanMS committed:
Update strings.json

RZomermanMS committed:
Update strings.json

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -14,8 +14,10 @@
+from homeassistant.exceptions import ServiceValidationError
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
+
```
```suggestion
```

joostlek commented on homeassistant/components/velbus/climate.py:
```diff
@@ -79,3 +85,16 @@ async def async_set_preset_mode(self, preset_mode: str) -> None:
+            )
+        await self._channel.set_mode(hvac_mode)
+        self.async_write_ha_state()
+
```
```suggestion
```

joostlek committed:
Apply suggestions from code review
